File: TIC-TAC-TOE.cpp
#include <iostream>
#include <vector>
#include <string>
#include <limits>
#include <cstdlib>
#include <ctime>

using namespace std;

class TicTacToe {
private:
    vector<char> board;

public:
    TicTacToe() : board(9, ' ') {}

    void printBoard() const {
        cout << "\n";
        cout << " " << board[0] << " | " << board[1] << " | " << board[2] << "\n";
        cout << "-----------\n";
        cout << " " << board[3] << " | " << board[4] << " | " << board[5] << "\n";
        cout << "-----------\n";
        cout << " " << board[6] << " | " << board[7] << " | " << board[8] << "\n";
        cout << "\n";
    }

    void playerMove(const string &playerName, char move) {
        int index;
        cout << "\nYour Move, " << playerName << "! Enter your index (1-9): ";
        while (true) {
            try {
                cin >> index;
                index--;
                if (cin.fail() || index < 0 || index >= 9 || board[index] != ' ') {
                    cin.clear();                                         // Clear the input stream state
                    cin.ignore(numeric_limits<streamsize>::max(), '\n'); // Clear the input buffer
                    cout << "\nInvalid input or index! Please enter a valid index (1-9): ";
                } else {
                    break;
                }
            } catch (...) {
                cout << "\nInvalid input or index! Please enter a valid index (1-9): ";
                cin.clear();                                         // Clear the input stream state
                cin.ignore(numeric_limits<streamsize>::max(), '\n'); // Clear the input buffer
            }
        }
        board[index] = move;
    }

    bool checkWinner() const {
        // Row Check
        for (int i = 0; i < 3; i++) {
            if (board[i] != ' ' && board[i] == board[i + 1] && board[i] == board[i + 2]) {
                return true;
            }
        }

        // Column Check
        for (int i = 0; i < 3; i++) {
            if (board[i] != ' ' && board[i] == board[i + 3] && board[i] == board[i + 6]) {
                return true;
            }
        }

        // Diagonal Check
        if (board[0] != ' ' && board[0] == board[4] && board[0] == board[8]) {
            return true;
        }
        if (board[2] != ' ' && board[2] == board[4] && board[2] == board[6]) {
            return true;
        }

        return false;
    }

    bool checkDraw() const {
        for (char cell : board) {
            if (cell == ' ') {
                return false;
            }
        }
        return true;
    }

    void playGame(const string &player1, const string &player2) {
        bool runningStatus = true;
        char currentPlayer = 'X';

        while (runningStatus) {
            printBoard();

            if (currentPlayer == 'X') {
                playerMove(player1, 'X');
            } else {
                playerMove(player2, 'O');
            }

            if (checkWinner()) {
                printBoard();
                cout << "Player " << currentPlayer << " (" << (currentPlayer == 'X' ? player1 : player2) << ") Wins!\n";
                runningStatus = false;
            } else if (checkDraw()) {
                printBoard();
                cout << "Game Draw!\n";
                runningStatus = false;
            }

            currentPlayer = (currentPlayer == 'X') ? 'O' : 'X';
        }
    }
};

int main() {
    TicTacToe game;
    string player1, player2;

    cout << "Enter Player 1 (X) Name: ";
    getline(cin, player1);

    cout << "Enter Player 2 (O) Name: ";
    getline(cin, player2);

    game.playGame(player1, player2);

    return 0;
}

File: TIC-TAC-TOE.py
#-------------------------------------------------------------
"""
| Name of author: Param Dholakia
| Topic: TIC-TAC-TOE
| Date: 15-04-2024
 """ 
#-------------------------------------------------------------


class TicTacToe:
    def __init__(self):
        # we will use a single list to represent the 3x3 board
        # For every empty cell we will use ' '.
        self.board = [' ' for _ in range(9)]        

    def printBoard(self):
        print()
        print(f" {self.board[0]} | {self.board[1]} | {self.board[2]} ")
        print("-----------")
        print(f" {self.board[3]} | {self.board[4]} | {self.board[5]} ")
        print("-----------")
        print(f" {self.board[6]} | {self.board[7]} | {self.board[8]} ")
        print()

    def playerMove(self, move, playerName):
        while True:
            try:
                index = int(input(f"\nYour Move, {playerName}! Enter your index (1-9): ")) - 1
                if 0 <= index < 9 and self.board[index] == ' ':
                    self.board[index] = move
                    break
                else:
                    print("Invalid index! Please enter a valid index.")
            except ValueError:
                print("Invalid input! Please enter a number.")
            
    def winnerCheck(self):
        for i in range(0, 9, 3):
            if self.board[i] != ' ' and self.board[i] == self.board[i + 1] == self.board[i + 2]:
                return True

        # Column Check
        for i in range(3):
            if self.board[i] != ' ' and self.board[i] == self.board[i + 3] == self.board[i + 6]:
                return True

        # Diagonal Check
        if self.board[0] != ' ' and self.board[0] == self.board[4] == self.board[8]:
            return True
        if self.board[2] != ' ' and self.board[2] == self.board[4] == self.board[6]:
            return True

        return False
    
    def check_draw(self):
        return ' ' not in self.board
    


if __name__ == '__main__':
    # Initialize the game
    TicTacToe = TicTacToe()
    
    print("Welcome to Tic-Tac-Toe!")
    player1 = input("Enter Player 1 (X) name: ")
    player2 = input("Enter Player 2 (O) name: ")

    runningCondition = True

    while runningCondition:
        TicTacToe.printBoard()
        TicTacToe.playerMove('X', player1)
        TicTacToe.printBoard()
        if TicTacToe.winnerCheck():
            print(f"Congratulations! {player1} wins!")
            runningCondition = False
            break
        if TicTacToe.check_draw():
            print("It's a draw!")
            runningCondition = False
            break

        TicTacToe.playerMove('O', player2)
        if TicTacToe.winnerCheck():
            TicTacToe.printBoard()
            print(f"Congratulations! {player2} wins!")
            runningCondition = False
            break
        if TicTacToe.check_draw():
            print("It's a draw!")
            runningCondition = False
            break
        